greenhouse_management/graphql/Device.py

```diff
@@ -0,0 +1,141 @@
+import graphene
+from graphene_django import DjangoObjectType
+from graphql_jwt.decorators import login_required
+from greenhouse_management.exceptions import PermissionDenied
+from greenhouse_management.graphql.GreenHouse import GreenHouseType
+from greenhouse_management.models import *
+
+
+class DeviceType(DjangoObjectType):
+    class Meta:
+        model = Device
+        fields = '__all__'
+    greenhouse = graphene.Field(GreenHouseType)
+
+    def resolve_greenhouse(self, info):
+        return self.greenhouse
+
+
+class FunctionalityEnum(graphene.Enum):
+    PASSIVE = 'PA'
+    ACTIVE = 'AC'
+
+
+class DeviceInput(graphene.InputObjectType):
+    name = graphene.String()
+    functionality = graphene.String()
+    greenhouse = graphene.ID()
+
+
+class CreateDevice(graphene.Mutation):
+    """To create device provide 'name', 'functionality' and 'greenhouse'."""
+
+    class Arguments:
+        input = DeviceInput(required=True)
+
+    device = graphene.Field(DeviceType)
+
+    @classmethod
+    @login_required
+    def mutate(cls, root, info, input):
+        request_user = info.context.user
+
+        try:         
+            if request_user.is_superuser:
+                greenhouse = GreenHouse.objects.get(id=input.greenhouse)
+            else:
+                greenhouse = GreenHouse.objects.get(id=input.greenhouse, owner=request_user)
+        except GreenHouse.DoesNotExist:
+            raise Exception("Greenhouse with this credentials does not exist.")
+
+        device = Device.objects.create(
+            name=input.name,
+            functionality=input.functionality,
+            greenhouse=greenhouse
+        )
+        device.save()
+
+        return CreateDevice(device=device)
+
+
+class UpdateDevice(graphene.Mutation):
+    """To update device change 'name', 'functionality' and 'greenhouse'."""
+
+    class Arguments:
+        input = DeviceInput(required=True)
+        id = graphene.Int(required=True)
+
+    device = graphene.Field(DeviceType)
+
+    @classmethod
+    @login_required
+    def mutate(cls, root, info, input, id):
+        device = Device.objects.get(pk=id)
+        if info.context.user.is_superuser or info.context.user == device.greenhouse.owner:
+            device.name = input.name if input.name not in ['', None] else device.name
+            device.functionality = input.functionality if input.functionality not in ['', None] else device.functionality
+            if input.greenhouse:
+                try:
+                    greenhouse = GreenHouse.objects.get(id=input.greenhouse)
```

Comment:
tu mozna dodac sprawdzenie czy info.context.user == greenhouse.owner

Reply:
Tu raczej nie trzeba tego dodawać bo w wyższym ifie w linii 74 jest sprawdzane czy info.context.user jest superuser lub czy jest właścicielem greenhouse więc jeśli nie jest ani tym ani tym to nie dojdzie do tego miejsca.

Reply:
naprawione